### PR TITLE
[WIP] Internal Storage Interface for Sterling Contracts

### PR DESCRIPTION
**Pull request by david-yan**

### CODE REVIEW

**Comment by nhynes** [comment]
can you rebase `eth-inf-gas` onto master and then rebase your pr onto the updated `eth-inf-gas`? there are a few commits here that don't belong. namely, those by Peter and Warren.

File: ethcore/src/externalities.rs [hunk 1]
```diff
@@ -373,11 +373,11 @@ impl<'a, T: 'a, V: 'a, B: 'a> Ext for Externalities<'a, T, V, B>
 	}
 
 	fn request_bytes(&mut self, key: H256) -> Option<Vec<u8>> {
-		self.storage.request_bytes(key)
+		Some(self.storage.request_bytes(key).unwrap())
```
Comment (nhynes) [inline ethcore/src/externalities.rs]: Why are you turning a `Result` into an `Option`? there's no scenario in which a valid read will yield no bytes

File: ethcore/src/externalities.rs [hunk 2]
```diff
@@ -373,11 +373,11 @@ impl<'a, T: 'a, V: 'a, B: 'a> Ext for Externalities<'a, T, V, B>
...
 	}
 
 	fn store_bytes(&mut self, key: H256, bytes: &[u8]) {
-		self.storage.store_bytes(key, bytes)
+		self.storage.store_bytes(key, bytes).unwrap()
```
Comment (nhynes) [inline ethcore/src/externalities.rs]: you should return `Result<()>` (i.e. don't unwrap)

File: ethcore/src/storage.rs [hunk 3]
```diff
@@ -1,18 +1,18 @@
 use ethereum_types::H256;
 
 pub trait Storage {
-	fn request_bytes(&mut self, key: H256) -> Option<Vec<u8>>;
-	fn store_bytes(&mut self, key: H256, bytes: &[u8]);
+	fn request_bytes(&mut self, key: H256) -> Result<Vec<u8>, ()>;
+	fn store_bytes(&mut self, key: H256, bytes: &[u8]) -> Result<(), ()>;
```
Comment (nhynes) [inline ethcore/src/storage.rs]: perhaps define a more descriptive error type? even a string would be more useful than Unit

File: ethcore/src/storage.rs [hunk 4]
```diff
@@ -1,18 +1,18 @@
...
 pub struct DummyStorage {}
 
 impl Storage for DummyStorage {
-	fn request_bytes(&mut self, key: H256) -> Option<Vec<u8>> {
+	fn request_bytes(&mut self, key: H256) -> Result<Vec<u8>, ()> {
 		unimplemented!();
 	}
 
-	fn store_bytes(&mut self, key: H256, bytes: &[u8]) {
+	fn store_bytes(&mut self, key: H256, bytes: &[u8]) -> Result<(), ()> {
```
Comment (nhynes) [inline ethcore/src/storage.rs]: more descriptive error type, please

File: ethcore/wasm/src/runtime.rs [hunk 5]
```diff
@@ -395,7 +395,7 @@ impl<'a> Runtime<'a> {
 					.map_err(|_| Error::BadUtf8)?;
 				match payload[0] {
 					1 => print!("{}", out_str),
-					2 => print!("error: {}", out_str),
+					2 => eprint!("{}", out_str),
```
Comment (nhynes) [inline ethcore/wasm/src/runtime.rs]: revert this change, please. `eprint` won't compile inside of the enclave

File: ethcore/src/externalities.rs (hunk 1 above)
Comment (david-yan) [reply ethcore/src/externalities.rs]: Sorry these changes weren't meant to be final or meant for review. I just wanted to get some version that compiled.

Comment (nhynes) [reply ethcore/src/externalities.rs]: gotcha; my bad. I'll just ignore this until the [WIP] is removed. thanks!

File: ethcore/src/machine.rs [hunk 6]
```diff
@@ -145,7 +146,8 @@ impl EthereumMachine {
 			call_type: CallType::Call,
 			params_type: ParamsType::Separate,
 		};
-		let mut ex = Executive::new(&mut state, &env_info, self);
+		let mut dummy_storage = DummyStorage {};
```
Comment (nhynes) [inline ethcore/src/machine.rs]: still want `DummyStorage::new()` even though it's a dummy :)

File: ethcore/src/spec/spec.rs [hunk 7]
```diff
@@ -638,7 +639,8 @@ impl Spec {
 				let mut substate = Substate::new();
 
 				{
-					let mut exec = Executive::new(&mut state, &env_info, self.engine.machine());
+					let mut dummy_storage = DummyStorage {};
```
Comment (nhynes) [inline ethcore/src/spec/spec.rs]: `DummyStorage::new()`

File: ethcore/src/state/mod.rs [hunk 8]
```diff
@@ -749,7 +750,8 @@ impl<B: Backend> State<B> {
 	fn execute<T, V>(&mut self, env_info: &EnvInfo, machine: &Machine, t: &SignedTransaction, options: TransactOptions<T, V>, virt: bool)
 		-> Result<Executed<T::Output, V::Output>, ExecutionError> where T: trace::Tracer, V: trace::VMTracer,
 	{
-		let mut e = Executive::new(self, env_info, machine);
+		let mut dummy_storage = DummyStorage {};
```
Comment (nhynes) [inline ethcore/src/state/mod.rs]: `DummyStorage::new()`

File: ethcore/vm/src/ext.rs [hunk 9]
```diff
@@ -148,4 +149,8 @@ pub trait Ext {
 
 	/// Check if running in static context.
 	fn is_static(&self) -> bool;
+
+	fn request_bytes(&mut self, key: H256) -> StdResult<Vec<u8>, String>;
```
Comment (nhynes) [inline ethcore/vm/src/ext.rs]: ehh, can  you actually make this use the parity `Result`? you might need to add some logic to `errors.rs`

File: ethcore/vm/src/tests.rs [hunk 10]
```diff
@@ -15,6 +15,7 @@
 // along with Parity.  If not, see <http://www.gnu.org/licenses/>.
 
 use std::sync::Arc;
+use std::result::Result as StdResult;
```
Comment (nhynes) [inline ethcore/vm/src/tests.rs]: no `StdResult`, please. let's keep error types consistent in the crate

File: ethcore/wasm/src/env.rs [hunk 11]
```diff
@@ -47,6 +47,8 @@ pub mod ids {
 	pub const SENDER_FUNC: usize = 190;
 	pub const ORIGIN_FUNC: usize = 200;
 	pub const ELOG_FUNC: usize = 210;
+	pub const REQUESTBYTES_FUNC: usize = 220;
```
Comment (nhynes) [inline ethcore/wasm/src/env.rs]: nit: `REQUEST_BYTES_FUNC`, `STORE_BYTES_FUNC`

File: ethcore/wasm/src/runtime.rs [hunk 12]
```diff
@@ -724,6 +725,26 @@ impl<'a> Runtime<'a> {
...
+
+		Ok(())
+	}
+
+	/// Signature: `fn store_bytes(bytes: *const u8, len: u64, key: *mut u8)`
+	pub fn store_bytes(&mut self, args: RuntimeArgs) -> Result<()> {
+		let bytes_ptr: u32 = args.nth_checked(0)?;
+		let len: u64 = args.nth_checked(1)?;
+		let bytes = self.memory.get(bytes_ptr, len as usize)?;
+		let key = self.ext.store_bytes(&bytes).unwrap();
```
Comment (nhynes) [inline ethcore/wasm/src/runtime.rs]: no unwrap. this should return an error

File: ethcore/wasm/src/runtime.rs [hunk 13]
```diff
@@ -724,6 +725,26 @@ impl<'a> Runtime<'a> {
...
+		self.memory.set(args.nth_checked(2)?, &*key);
```
Comment (nhynes) [inline ethcore/wasm/src/runtime.rs]: clever use of `AsRef` :)

File: ethcore/vm/src/ext.rs (hunk 9 above)
Comment (david-yan) [reply ethcore/vm/src/ext.rs]: I thought about using the parity `Result` but the `Error` provided is a VM error and I wasn't sure whether an error accessing the storage would be considered a VM error.

Comment (david-yan) [reply ethcore/vm/src/ext.rs]: If it should be considered a VM error I can add an option to the enum for `Storage(string)`?

Comment (nhynes) [reply ethcore/vm/src/ext.rs]: it's on the same level of error as https://github.com/ekiden/parity/blob/master/ethcore/vm/src/ext.rs#L115

File: ethcore/wasm/src/runtime.rs (hunk 12 above)
Comment (david-yan) [reply ethcore/wasm/src/runtime.rs]: fixed

File: ethcore/wasm/src/env.rs (hunk 11 above)
Comment (david-yan) [reply ethcore/wasm/src/env.rs]: fixed

File: ethcore/vm/src/ext.rs (hunk 9 above)
Comment (david-yan) [reply ethcore/vm/src/ext.rs]: fixed

File: ethcore/src/machine.rs (hunk 6 above)
Comment (david-yan) [reply ethcore/src/machine.rs]: fixed for all instances
